tests/integration/suite/daprd/pubsub/client/basic.go

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2023 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implieh.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	pubsub "github.com/dapr/dapr/pkg/api/clients/pubsub"
+	"github.com/dapr/dapr/tests/integration/framework"
+	procdaprd "github.com/dapr/dapr/tests/integration/framework/process/daprd"
+	"github.com/dapr/dapr/tests/integration/suite"
+)
+
+func init() {
+	suite.Register(new(basic))
+}
+
+type basic struct {
+	daprd *procdaprd.Daprd
+}
+
+func (b *basic) Setup(t *testing.T) []framework.Option {
+	b.daprd = procdaprd.New(t, procdaprd.WithResourceFiles(`
+apiVersion: dapr.io/v1alpha1
+kind: Component
+metadata:
+  name: mypubsub
+spec:
+  type: pubsub.in-memory
+  version: v1
+`))
+
+	return []framework.Option{
+		framework.WithProcesses(b.daprd),
+	}
+}
+
+func (b *basic) Run(t *testing.T, ctx context.Context) {
+	b.daprd.WaitUntilRunning(t, ctx)
+	daprdHTTPURL := fmt.Sprintf("http://localhost:%d/", b.daprd.HTTPPort())
+	client, err := pubsub.NewClient(daprdHTTPURL)
+	pubsubName := "mypubsub"
+	topicName := "mytopic"
+	if err != nil {
+		panic(err)
+	}
+
+	t.Run("good json", func(t *testing.T) {
+		for _, body := range []string{
+			"[]",
+			`[{"key": "key1", "vae": "value1"}]`,
+			`[{"key": "key1", "value": "value1"}]`,
+			`[{"key": "key1", "value": "value1"},{"key": "key2", "value": "value1"}]`,
+			`[{"key": "key1", "value": "value1"},{"key": "key2", "value": "value1"},  {"key": "key1", "value": "value1"},{"key": "key2", "value": "value1"}]`,
+		} {
+			t.Run(body, func(t *testing.T) {
+				resp, err := client.PostPublishMessageWithBody(ctx, pubsubName, topicName, "application/json", strings.NewReader(body))
```

Comment:
I would this and all API client functions to accept a struct, rather than a set of bare parameters. Bare parameters introduce ambiguity (it's particularly bad here with 3 consecutive strings), restrict our ability to extend the API in future in a backwards compatible way, and generally doesn't follow how other client packages work, like Go's `http.Client` and `.Do()`.

Reply:
@JoshVanL that's a good point but I am following the style being used by the SDKs, this is something that we might want to improve for all SDKs at some point.

Reply:
@yaron2 @artursouza I need your input here..

Reply:
I agree with Josh. What we do in the Java SDK is to have a "catch all" request type and then overload the method with shorter versions with combinations of up to 5 params.